### PR TITLE
fix: add an event when tracks are published with no data

### PR DESCRIPTION
**Pull request by raviteja83**
### Details(context, link the issue, how was the bug fixed, what does the new feature do)

- Handle possible scenarios where this can be recovered
- Happens when whatsapp call is already happening and 100ms link is joined on iOS mweb.

### CODE REVIEW

File: packages/hms-video-store/src/sdk/index.ts [hunk 1]
```diff
@@ -1562,12 +1567,10 @@ export class HMSSdk implements HMSInterface {
...
-    // this.sendAnalyticsEvent(
-    //   AnalyticsEventFactory.audioDetectionFail(error, this.deviceManager.getCurrentSelection().audioInput),
-    // );
-    if (this.isDiagnostics) {
-      this.listener?.onError(error);
-    }
+    this.sendAnalyticsEvent(
+      AnalyticsEventFactory.audioDetectionFail(error, this.deviceManager.getCurrentSelection().audioInput),
+    );
+    this.listener?.onError(error);
```
Comment (raviteja83) [inline packages/hms-video-store/src/sdk/index.ts:1573]: there are false positives for this. can we check the dataArray values in the track level monitor? how long are they 128? @eswarclynn can you help here

File: packages/hms-video-store/src/media/tracks/HMSTrack.ts [hunk 2]
```diff
@@ -85,6 +85,11 @@ export abstract class HMSTrack {
   protected setFirstTrackId(trackId: string) {
     this.firstTrackId = trackId;
   }
+
+  isTrackNotPublishing = () => {
+    return this.nativeTrack.readyState === 'ended' || this.nativeTrack.muted;
```
Comment (hdz-666) [inline packages/hms-video-store/src/media/tracks/HMSTrack.ts:90]: should be enum

File: packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts [hunk 3]
```diff
@@ -151,16 +148,19 @@ export class HMSLocalAudioTrack extends HMSAudioTrack {
...
     try {
       const newTrack = await getAudioTrack(settings);
+      this.addTrackEventListeners(newTrack);
       this.tracksCreated.add(newTrack);
       HMSLogger.d(this.TAG, 'replaceTrack, Previous track stopped', prevTrack, 'newTrack', newTrack);
       await this.updateTrack(newTrack);
     } catch (e) {
       // Generate a new track from previous settings so there will be audio because previous track is stopped
       const newTrack = await getAudioTrack(this.settings);
+      this.addTrackEventListeners(newTrack);
```
Comment (hdz-666) [inline packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:163]: why are we doing this twice?

Comment (raviteja83) [reply packages/hms-video-store/src/media/tracks/HMSLocalAudioTrack.ts:163]: needs everywhere a new track is created/ this is a fallback where in some cases, deviceId or some other settings fails.

File: packages/hms-video-store/src/media/tracks/HMSTrack.ts (hunk 2 above)
Comment (raviteja83) [reply packages/hms-video-store/src/media/tracks/HMSTrack.ts:90]: not our type. so not planning to add, as it is populated by autocomplete

File: packages/hms-video-store/src/error/ErrorCodes.ts [hunk 4]
```diff
@@ -76,6 +76,9 @@ export const ErrorCodes = {
 
     // Selected device not detected on change
     SELECTED_DEVICE_MISSING: 3014,
+
+    // Track is publishing with no data, can happen when a whatsapp call is ongoing before 100ms call in mweb
+    NO_DATA: 3015,
```
[the remark this answers is not in the document]
Comment (raviteja83) [reply packages/hms-video-store/src/error/ErrorCodes.ts:81]: should be sufficient i feel.

**Review by ygit — DISMISSED** [review]
LGTM

**Review by ygit — APPROVED** [review]

**Review by KaustubhKumar05 — APPROVED** [review]

**Pull request merged**